core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -216,6 +227,11 @@ public ActionRequestValidationException validate() {
             validationException = addValidationError("id is too long, must be no longer than 512 bytes but was: " +
                             id.getBytes(StandardCharsets.UTF_8).length, validationException);
         }
+
+        if (id == null && (versionType == VersionType.INTERNAL && version == Versions.MATCH_ANY) == false) {
```

Comment:
can we put an extra validation that's explicit for the opType check - i.e., `if (opType() != OpType.INDEX && id == 1) addValidationError("an id is required for a " + opType() + " operation")` (preferably next the current ` if (opType() == OpType.CREATE) { ` line). I think people will be confused by this message if they use create and we should also catch the delete case, while at it.